vc/vcjwt.go

```diff
@@ -0,0 +1,149 @@
+// While the jwt claims should match the counterpart values in the vc model, it's possible that they don't
+// but there would be no way to know if they don't match given that they're overwritten.
+//
+// [spec]: https://www.w3.org/TR/vc-data-model/#json-web-token
+func Decode[T CredentialSubject](vcJWT string) (DecodedVCJWT[T], error) {
+	decoded, err := jwt.Decode(vcJWT)
+	if err != nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("failed to decode vc-jwt: %w", err)
+	}
+
+	if decoded.Claims.Misc == nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("vc-jwt missing vc claim")
+	}
+
+	if _, ok := decoded.Claims.Misc["vc"]; ok == false {
+		return DecodedVCJWT[T]{}, fmt.Errorf("vc-jwt missing vc claim")
+	}
+
+	bytes, err := json.Marshal(decoded.Claims.Misc["vc"])
+	if err != nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("failed to decode vc claim: %w", err)
+	}
+
+	var vc DataModel[T]
+	if err := json.Unmarshal(bytes, &vc); err != nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("failed to decode vc claim: %w", err)
+	}
+
+	// the following conditionals are included to conform with the jwt decoding section
+	// of the specification defined here: https://www.w3.org/TR/vc-data-model/#jwt-decoding
+	if decoded.Claims.Issuer != "" {
+		vc.Issuer = decoded.Claims.Issuer
+	}
+
+	if decoded.Claims.JTI != "" {
+		vc.ID = decoded.Claims.JTI
+	}
+
+	if decoded.Claims.Subject != "" {
+		vc.CredentialSubject.SetID(decoded.Claims.Subject)
+	}
+
+	if decoded.Claims.Expiration != 0 {
+		vc.ExpirationDate = time.Unix(decoded.Claims.Expiration, 0).UTC().Format(time.RFC3339)
+	}
+
+	if decoded.Claims.NotBefore != 0 {
+		vc.IssuanceDate = time.Unix(decoded.Claims.NotBefore, 0).UTC().Format(time.RFC3339)
+	}
+
+	return DecodedVCJWT[T]{
+		JWT: decoded,
+		VC:  vc,
+	}, nil
+
+}
+
+// DecodedVCJWT represents a decoded vc-jwt. It contains the decoded jwt and decoded vc data model
+type DecodedVCJWT[T CredentialSubject] struct {
+	JWT jwt.Decoded
+	VC  DataModel[T]
+}
+
+// Verify verifies the decoded vc-jwt. It checks for the presence of required fields and verifies the jwt.
+func (vcjwt DecodedVCJWT[T]) Verify() error {
+	if vcjwt.VC.Issuer == "" {
+		return errors.New("verification failed. missing issuer")
+	}
+
+	if vcjwt.VC.ID == "" {
+		return errors.New("verification failed. missing id")
+	}
+
+	if vcjwt.VC.IssuanceDate == "" {
+		return errors.New("verification failed. missing issuance date")
+	}
+
+	issuanceDate, err := time.Parse(time.RFC3339, vcjwt.VC.IssuanceDate)
+	if err != nil {
+		return fmt.Errorf("verification failed. failed to parse issuance date: %w", err)
+	}
+
+	if time.Now().UTC().Before(issuanceDate.UTC()) {
+		return fmt.Errorf("verification failed. vc cannot be used before %s", vcjwt.VC.IssuanceDate)
+	}
+
+	if vcjwt.VC.ExpirationDate != "" {
+		exp, err := time.Parse(time.RFC3339, vcjwt.VC.ExpirationDate)
+		if err != nil {
+			return fmt.Errorf("verification failed. failed to parse expiration date: %w", err)
```

Comment:
good call and agreed full-stop makes it kinda weird. debating between `;` and `-`

`verification failed - failed to parse expiration date`  
orrr
`verification failed; failed to parse expiration date`'

actually thinking about it a bit more, it may make the most sense to remove all the `verification failed`s entirely because the usage pattern actually makes it clear that something about verification failed. e.g.

```go
decoded, err := cred.Verify()
if err != nil {
    // handle error
}
```

looking at ^ makes `"verification failed"` seem a bit redundant. caller can always wrap the error and add more context if needed

Reply:
Yes, great point and agreed.